spec/unit/Java.spec.js

```diff
@@ -0,0 +1,158 @@
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+*/
+
+const path = require('path');
+const rewire = require('rewire');
+const { CordovaError } = require('cordova-common');
+const utils = require('../../bin/templates/cordova/lib/utils');
+
+describe('Java', () => {
+    const Java = rewire('../../bin/templates/cordova/lib/Java');
+    let originalEnv = null;
+
+    beforeAll(() => {
+        originalEnv = Object.assign({}, process.env);
+    });
+
+    afterAll(() => {
+        process.env = Object.assign({}, originalEnv);
+    });
+
+    describe('getVersion', () => {
+        beforeEach(() => {
+            process.env = Object.assign({}, originalEnv);
+        });
+
+        it('runs', async () => {
+            Java.__set__('execa', () => Promise.resolve({
+                all: 'javac 1.8.0_275'
+            }));
+
+            const versionObj = {
+                major: 1,
+                minor: 8,
+                patch: 0,
+                build: 275,
+                string: '1.8.0_275'
+            };
+
+            await expectAsync(Java.getVersion()).toBeResolvedTo(versionObj);
+        });
+
+        it('produces a CordovaError on error', async () => {
+            Java.__set__('execa', () => Promise.reject({
+                shortMessage: 'test error'
+            }));
+            const emitSpy = jasmine.createSpy('events.emit');
+            Java.__set__('events', {
+                emit: emitSpy
+            });
+
+            await expectAsync(Java.getVersion()).toBeRejectedWithError(CordovaError, /Failed to run "javac -version"/);
+            expect(emitSpy).toHaveBeenCalledWith('verbose', 'test error');
+        });
+    });
+
+    describe('_ensure', () => {
+        beforeEach(() => {
+            process.env = Object.assign({}, originalEnv);
+            Java.__set__('javaIsEnsured', false);
+        });
+
+        it('with JAVA_HOME / without javac', async () => {
+            spyOn(utils, 'forgivingWhichSync').and.returnValue('');
+            process.env.JAVA_HOME = '/tmp/jdk';
+
+            await Java._ensure();
+
+            expect(process.env.PATH.split(path.delimiter))
+                .toContain(path.join(process.env.JAVA_HOME, 'bin'));
+        });
+
+        it('without JAVA_HOME / with javac - Mac OS X - success', async () => {
+            spyOn(utils, 'forgivingWhichSync').and.returnValue('/tmp/jdk/bin');
+            process.env.JAVA_HOME = '';
+            const fsSpy = jasmine.createSpy('fs').and.returnValue(true);
+            Java.__set__('fs', {
+                existsSync: fsSpy
+            });
+            Java.__set__('execa', jasmine.createSpy('execa').and.returnValue(Promise.resolve({
+                stdout: '/tmp/jdk'
+            })));
```

Comment:
If we do not use the features of Jasmine spies, I would prefer using plain JS function as test-fakes. It makes for far less verbose code IMHO.
```suggestion
            Java.__set__('execa', async () => ({ stdout: '/tmp/jdk' }));
```

This might be a matter of taste of course.